source/common/http/http1/parser_impl.cc

```diff
@@ -0,0 +1,226 @@
+#include "common/http/http1/parser_impl.h"
+
+#include <llhttp.h>
+
+#include <iostream>
+
+#include "common/common/assert.h"
+#include "common/http/http1/parser.h"
+
+namespace Envoy {
+namespace Http {
+namespace Http1 {
+namespace {
+ParserStatus intToStatus(int rc) {
+  // See
+  // https://github.com/nodejs/http-parser/blob/5c5b3ac62662736de9e71640a8dc16da45b32503/http_parser.h#L72.
+  switch (rc) {
+  case -1:
+    return ParserStatus::Error;
+  case 0:
+    return ParserStatus::Success;
+  case 1:
+    return ParserStatus::NoBody;
+  case 2:
+    return ParserStatus::NoBodyData;
+  case 21:
+    return ParserStatus::Paused;
+  default:
+    return ParserStatus::Unknown;
+  }
+}
+} // namespace
+
+class HttpParserImpl::Impl {
+public:
+  Impl(llhttp_type_t type) {
+    llhttp_init(&parser_, type, &settings_);
+    llhttp_set_lenient_chunked_length(&parser_, 1);
```

Comment:
I could pipe https://github.com/envoyproxy/envoy/blob/c93d486c2d303dbf2cc9f88717de0bfd66e1afdb/include/envoy/http/codec.h#L438 in to the parser. Like the lenient header validation, it would change the response code/message. WDYT?

Reply:
What code in Envoy is generating the error response when both transfer-encoding: chunked and content-length is set?

Reply:
https://github.com/envoyproxy/envoy/blob/c93d486c2d303dbf2cc9f88717de0bfd66e1afdb/source/common/http/http1/codec_impl.cc#L750-L753

Reply:
Line 718 already circumvents this test.

Reply:
The test at 747 looks altogether incorrect, whether chunked is disallowed doesn't impact the rule of 'never both'

Reply:
Why does 718 circumvent this block?

`codec_settings_.allow_chunked_length_`: allows Envoy to process requests with both C-L and TE without rejecting (and remove C-L). So I think it does impact the rule of never both..

Reply:
So, it should be read as allow_chunked_and_content_length_ ... that is sensible (very unwise, but not a default nor a defect)

Reply:
And the same toggle check should probably occur at 720.

Reply:
I'm not sure I understand why. That's removing a meaningless 0 content-length and otherwise rejecting the request since we shouldn't have a payload for CONNECT, not because TE: chunked is present.

Reply:
It seems to me that the integration tests pass with the current code.  I'm curious about which fail if we were to flip the value of these llhttp library settings.

Reply:
So long as we flipped based on the codec setting, the only change would be the status messages/response details:
```
Expected equality of these values:
  status.message()
    Which is: "http/1.1 protocol error: HPE_UNEXPECTED_CONTENT_LENGTH"
  "http/1.1 protocol error: both 'Content-Length' and 'Transfer-Encoding' are set."
    Which is: 0x3a24fa
```

I think adding a comment that our codec code handling this could be removed after clean up would make sense.